### ISSUE TITLE
numpy.ndarray messed up after unpickling (migrated from Trac #551)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#1158]
Original ticket http://projects.scipy.org/numpy/ticket/551
Reported 2007-07-16 by trac user cotackst, assigned to unknown.

After unpickling a pickled numpy.ndarray matrix (using cPickle), I get a segmentation fault when trying to multiply the matrix by a vector. However I can multiply a copy of the matrix with the same vector.

The code to generate the actual matrix is quite large, so I only attach the pickled matrix ('mat.bug').

Reproducing code:[[BR]]
from cPickle import Unpickler[[BR]]
from numpy import ones, dot[[BR]]

bug_mat = Unpickler(file('mat.bug','r')).load()[[BR]]
ok_mat = bug_mat.copy()[[BR]]
v = ones((1,bug_mat.shape[0]))[[BR]]
dot(v,ok_mat) # this works[[BR]]
dot(v,bug_mat) # this segfault[[BR]]

**Comment by thouis** [comment]
Attachment in Trac by trac user cotackst, 2007-07-16: [mat.bug](http://projects.scipy.org/numpy/attachment/ticket/551/mat.bug)

**Comment by thouis** [comment]
Comment in Trac by atmention:alberts, 2007-07-17

I think that if you could reduce the code to generate the matrix to the simplest case that causes an invalid matrix, that would be very helpful. Alternatively, attach the complete code here and someone else can take a stab at reducing it to a smaller test case.

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2007-09-21

I cannot reproduce the problem.

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2007-09-30

I can confirm the problem exists.

```
#9  0xb768747a in ATL_dgemvN_a1_x1_b0_y1 () from /usr/lib/sse2/libf77blas.so.3
```

Numpy r4070, Python 2.5, Linux 32.

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2007-09-30

Question: after unpickling, the flags look like this

```
  OWNDATA : False
  WRITEABLE : True
```

Why doesn't the array own its data, and why is it then writeable?

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2007-10-24

Minimal test case:

```
import cPickle
import numpy as N

data = N.random.random((3,3))
cPickle.dump(data,file('mat2.bug','w'))

bug_mat = cPickle.load(file('mat2.bug'))
ok_mat = bug_mat.copy()

v = N.ones((1,bug_mat.shape[0]))
N.dot(v,ok_mat) # this works
N.dot(v,bug_mat) # this segfault
```

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2007-10-24

I still cannot reproduce this problem (v4209 of NumPy).  

The array does not own its own data so that on read a copy does not need to be made (the object that read the pickle owns the data).  Yes, you can write to it. This is expected behavior.

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2007-10-24

I am running Numpy r4276.

```
blas_opt_info:
    libraries = ['f77blas', 'cblas', 'atlas']
    library_dirs = ['/usr/lib/sse2']
    define_macros = [('ATLAS_INFO', '"\\"3.6.0\\""')]
    language = c
    include_dirs = ['/usr/include']

lapack_opt_info:
    libraries = ['lapack', 'f77blas', 'cblas', 'atlas']
    library_dirs = ['/usr/lib']
    define_macros = [('ATLAS_INFO', '"\\"3.6.0\\""')]
    language = f77
    include_dirs = ['/usr/include']

atlas_info:
    libraries = ['lapack', 'f77blas', 'cblas', 'atlas']
    library_dirs = ['/usr/lib']
    language = f77
    include_dirs = ['/usr/include']

atlas_blas_info:
    libraries = ['f77blas', 'cblas', 'atlas']
    library_dirs = ['/usr/lib/sse2']
    language = c
    include_dirs = ['/usr/include']
```

I'll try a bit later without SSE2 atlas.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2007-11-03

I can't reproduce this either. Numpy 1.0.4.dev4338, Fedora8test3, python 2.5.1, self compiled ATLAS.

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2007-12-05

I can reproduce this bug on Ubuntu 7.10 -shipped python2.5 and SSE2-Atlas 3.6.0-20.6 (atlas3-sse2 package), and self-compiled Numpy 1.0.4.

However, with Numpy compiled without Atlas, the bug is not present.

With Numpy compiled with Atlas without processor-specific optimizations (atlas-base package), the bug is also not present.

With Numpy compiled with SSE-Atlas (atlas3-sse package), the bug is also not present.

So, this may be associated with some properties of SSE2-enabled Atlas, possibly the one shipped with Ubuntu/others. However, it is quite strange that this crash appears only with unpickled arrays --- the data pointer probably is valid as the array prints/copies fine, and the original and the unpickled arrays have the same strides.

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2007-12-20

This seems to be the same as [http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=448530 Debian bug #448530].

**Comment by thouis** [comment]
Comment in Trac by trac user bolme, 2008-02-08

I was able to reproduce this problem in MacOS X Leopard and Ubuntu 7.04.  I ran valgrind under Ubuntu.  Valgrind suggests the error may be caused by uninitialized values.

Test code...
    import pickle
    from numpy import array,dot,ones

```
# Create the arrays
a = ones((47,128*128))
b = ones((128*128,2))
c = ones((128*128,1))

# Test dot
print "Computing dot product on original arrays..."
print dot(a,b).shape # WORKS
print dot(a,c).shape # WORKS

print "Pickle and unpickel the array with workaround."
buffer = pickle.dumps(a)
a = array(pickle.loads(buffer))

# Test dot
print "Computing dot product with workaround..."
print dot(a,b).shape # WORKS
print dot(a,c).shape # WORKS

print "Pickle and unpickel the array without workaround."
a = pickle.loads(buffer)

# Test dot
print "Computing dot product without workaround..."
print dot(a,b).shape # WORKS
print dot(a,c).shape # CRASH 
```

Valgrind output:

```
Computing dot product on original arrays...
==13157== 
==13157== Conditional jump or move depends on uninitialised value(s)
==13157==    at 0x48FCA42: ATL_dmmK (in /home/bolme/tools/lib/python2.6/site-packages/numpy/core/_dotblas.so)
==13157== 
==13157== Use of uninitialised value of size 4
==13157==    at 0x48794C0: ATL_dgecopy (in /home/bolme/tools/lib/python2.6/site-packages/numpy/core/_dotblas.so)
==13157== 
==13157== Use of uninitialised value of size 4
==13157==    at 0x48794CE: ATL_dgecopy (in /home/bolme/tools/lib/python2.6/site-packages/numpy/core/_dotblas.so)
==13157== 
==13157== Use of uninitialised value of size 4
==13157==    at 0x4879531: ATL_dgecopy (in /home/bolme/tools/lib/python2.6/site-packages/numpy/core/_dotblas.so)
==13157== 
==13157== Conditional jump or move depends on uninitialised value(s)
==13157==    at 0x48FCBAF: ATL_dmmK (in /home/bolme/tools/lib/python2.6/site-packages/numpy/core/_dotblas.so)
(47, 2)
(47, 1)
Pickle and unpickel the array with workaround.
Computing dot product with workaround...
(47, 2)
(47, 1)
Pickle and unpickel the array without workaround.
Computing dot product without workaround...
(47, 2)
(47, 1)
==13157== 
==13157== ERROR SUMMARY: 41065 errors from 10 contexts (suppressed: 2174 from 4)
==13157== 
==13157== 1 errors in context 1 of 10:
==13157== Invalid read of size 4
==13157==    at 0x4014759: (within /lib/ld-2.5.so)
==13157==    by 0x4005B69: (within /lib/ld-2.5.so)
==13157==    by 0x4007995: (within /lib/ld-2.5.so)
==13157==    by 0x400B836: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x400BA2A: (within /lib/ld-2.5.so)
==13157==    by 0x4010DF4: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x40108ED: (within /lib/ld-2.5.so)
==13157==    by 0x4189C2C: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x418A2AB: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==  Address 0x46AD6BC is 36 bytes inside a block of size 38 alloc'd
==13157==    at 0x4021620: malloc (vg_replace_malloc.c:149)
==13157==    by 0x4007F33: (within /lib/ld-2.5.so)
==13157==    by 0x400B836: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x400BA2A: (within /lib/ld-2.5.so)
==13157==    by 0x4010DF4: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x40108ED: (within /lib/ld-2.5.so)
==13157==    by 0x4189C2C: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x418A2AB: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x4189B63: dlopen (in /lib/tls/i686/cmov/libdl-2.5.so)
==13157== 
==13157== 3 errors in context 2 of 10:
==13157== Conditional jump or move depends on uninitialised value(s)
==13157==    at 0x48FCBAF: ATL_dmmK (in /home/bolme/tools/lib/python2.6/site-packages/numpy/core/_dotblas.so)
==13157== 
==13157== 3 errors in context 3 of 10:
==13157== Invalid read of size 4
==13157==    at 0x4014743: (within /lib/ld-2.5.so)
==13157==    by 0x4005B69: (within /lib/ld-2.5.so)
==13157==    by 0x4007995: (within /lib/ld-2.5.so)
==13157==    by 0x400B836: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x400BA2A: (within /lib/ld-2.5.so)
==13157==    by 0x4010DF4: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x40108ED: (within /lib/ld-2.5.so)
==13157==    by 0x4189C2C: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x418A2AB: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==  Address 0x45AE548 is 16 bytes inside a block of size 19 alloc'd
==13157==    at 0x4021620: malloc (vg_replace_malloc.c:149)
==13157==    by 0x4007F33: (within /lib/ld-2.5.so)
==13157==    by 0x400B836: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x400BA2A: (within /lib/ld-2.5.so)
==13157==    by 0x4010DF4: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x40108ED: (within /lib/ld-2.5.so)
==13157==    by 0x4189C2C: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x418A2AB: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x4189B63: dlopen (in /lib/tls/i686/cmov/libdl-2.5.so)
==13157== 
==13157== 3 errors in context 4 of 10:
==13157== Invalid read of size 4
==13157==    at 0x4014743: (within /lib/ld-2.5.so)
==13157==    by 0x4005B69: (within /lib/ld-2.5.so)
==13157==    by 0x4007995: (within /lib/ld-2.5.so)
==13157==    by 0x4010D94: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x40108ED: (within /lib/ld-2.5.so)
==13157==    by 0x4189C2C: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x418A2AB: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x4189B63: dlopen (in /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x40FCB76: _PyImport_GetDynLoadFunc (dynload_shlib.c:130)
==13157==    by 0x40EAB17: _PyImport_LoadDynamicModule (importdl.c:42)
==13157==  Address 0x4629440 is 64 bytes inside a block of size 66 alloc'd
==13157==    at 0x4021620: malloc (vg_replace_malloc.c:149)
==13157==    by 0x4006EB4: (within /lib/ld-2.5.so)
==13157==    by 0x40078D9: (within /lib/ld-2.5.so)
==13157==    by 0x4010D94: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x40108ED: (within /lib/ld-2.5.so)
==13157==    by 0x4189C2C: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x418A2AB: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x4189B63: dlopen (in /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x40FCB76: _PyImport_GetDynLoadFunc (dynload_shlib.c:130)
==13157==    by 0x40EAB17: _PyImport_LoadDynamicModule (importdl.c:42)
==13157== 
==13157== 4 errors in context 5 of 10:
==13157== Invalid read of size 4
==13157==    at 0x4014770: (within /lib/ld-2.5.so)
==13157==    by 0x4005B69: (within /lib/ld-2.5.so)
==13157==    by 0x4007995: (within /lib/ld-2.5.so)
==13157==    by 0x4010D94: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x40108ED: (within /lib/ld-2.5.so)
==13157==    by 0x4189C2C: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x418A2AB: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x4189B63: dlopen (in /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x40FCB76: _PyImport_GetDynLoadFunc (dynload_shlib.c:130)
==13157==    by 0x40EAB17: _PyImport_LoadDynamicModule (importdl.c:42)
==13157==  Address 0x44F65E0 is 56 bytes inside a block of size 57 alloc'd
==13157==    at 0x4021620: malloc (vg_replace_malloc.c:149)
==13157==    by 0x4006EB4: (within /lib/ld-2.5.so)
==13157==    by 0x40078D9: (within /lib/ld-2.5.so)
==13157==    by 0x4010D94: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x40108ED: (within /lib/ld-2.5.so)
==13157==    by 0x4189C2C: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x418A2AB: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x4189B63: dlopen (in /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x40FCB76: _PyImport_GetDynLoadFunc (dynload_shlib.c:130)
==13157==    by 0x40EAB17: _PyImport_LoadDynamicModule (importdl.c:42)
==13157== 
==13157== 14 errors in context 6 of 10:
==13157== Invalid read of size 4
==13157==    at 0x4014759: (within /lib/ld-2.5.so)
==13157==    by 0x4005B69: (within /lib/ld-2.5.so)
==13157==    by 0x4007995: (within /lib/ld-2.5.so)
==13157==    by 0x4010D94: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x40108ED: (within /lib/ld-2.5.so)
==13157==    by 0x4189C2C: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x418A2AB: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x4189B63: dlopen (in /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x40FCB76: _PyImport_GetDynLoadFunc (dynload_shlib.c:130)
==13157==    by 0x40EAB17: _PyImport_LoadDynamicModule (importdl.c:42)
==13157==  Address 0x44A602C is 52 bytes inside a block of size 55 alloc'd
==13157==    at 0x4021620: malloc (vg_replace_malloc.c:149)
==13157==    by 0x4006EB4: (within /lib/ld-2.5.so)
==13157==    by 0x40078D9: (within /lib/ld-2.5.so)
==13157==    by 0x4010D94: (within /lib/ld-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x40108ED: (within /lib/ld-2.5.so)
==13157==    by 0x4189C2C: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x400CFA5: (within /lib/ld-2.5.so)
==13157==    by 0x418A2AB: (within /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x4189B63: dlopen (in /lib/tls/i686/cmov/libdl-2.5.so)
==13157==    by 0x40FCB76: _PyImport_GetDynLoadFunc (dynload_shlib.c:130)
==13157==    by 0x40EAB17: _PyImport_LoadDynamicModule (importdl.c:42)
==13157== 
==13157== 852 errors in context 7 of 10:
==13157== Conditional jump or move depends on uninitialised value(s)
==13157==    at 0x48FCA42: ATL_dmmK (in /home/bolme/tools/lib/python2.6/site-packages/numpy/core/_dotblas.so)
==13157== 
==13157== 855 errors in context 8 of 10:
==13157== Use of uninitialised value of size 4
==13157==    at 0x4879531: ATL_dgecopy (in /home/bolme/tools/lib/python2.6/site-packages/numpy/core/_dotblas.so)
==13157== 
==13157== 19665 errors in context 9 of 10:
==13157== Use of uninitialised value of size 4
==13157==    at 0x48794CE: ATL_dgecopy (in /home/bolme/tools/lib/python2.6/site-packages/numpy/core/_dotblas.so)
==13157== 
==13157== 19665 errors in context 10 of 10:
==13157== Use of uninitialised value of size 4
==13157==    at 0x48794C0: ATL_dgecopy (in /home/bolme/tools/lib/python2.6/site-packages/numpy/core/_dotblas.so)
```

The errors with ATL_dmmK and ATL_dgecopy only seem to occure with the line that typically causes the segfault.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2008-02-24

Should we close this ticket? It looks like an ATLAS problem.

**Comment by thouis** [comment]
Comment in Trac by trac user bolme, 2008-02-25

I am not convinced that this is an atlas problem.  I think it may be an issue with initialization of the ndarray during the pickle process.  The output of valgrind indicates initialized values.  How does the typical constructor differ than the one used for pickling?

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2008-02-25

The only real difference is that the memory for the array on unpickling is allocated originally as a Python String (because that is what unpickling does).  NumPy uses that memory directly instead of copying it to another memory area. 

Of course, perhaps something is not set correctly in the array in the **setitem** method and this could be checked again (look in arraymethods.c file).

**Comment by thouis** [comment]
Comment in Trac by atmention:rkern, 2008-02-25

Replying to [comment:14 oliphant]:

> The only real difference is that the memory for the array on unpickling is allocated originally as a Python String (because that is what unpickling does).  !NumPy uses that memory directly instead of copying it to another memory area. 

Do we keep a reference to that string object? If not, what happens to the memory when the string's refcount hits 0?

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2008-03-11

**Comment by thouis** [comment]
Comment in Trac by atmention:jarrodmillman, 2008-04-09

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2008-04-10

Can this be an alignment issue?
    import numpy, cPickle as pickle
    x = numpy.zeros((3,4))
    y = pickle.loads(pickle.dumps(x))
    v = numpy.ones((1, y.shape[0]))

```
print "x:", x.__array_interface__['data'][0] % x.dtype.itemsize
print "y:", y.__array_interface__['data'][0] % y.dtype.itemsize

numpy.dot(v, x)
numpy.dot(v, y)
```

outputs
    x: 0
    y: 4
    Segmentation fault

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2008-04-10

Any comment on Robert's question?

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2008-04-10

I think a reference to the unpickled string is kept in self->base, and to my understanding it is INCREF'd properly on arraymethods.c:1303. But I'm not familiar with this code, so a more informed opinion would be helpful. But note that replacing in the code above
    print "y:", y.**array_interface**['data'][0] % y.dtype.itemsize, sys.getrefcount(y.base), type(y.base)
I get
    y: 4 2 <type 'str'>
So the base of y is a string with refcount 2, which is ok (but it's not aligned at double-boundary). Also, if it was a reference count problem, it would be difficult to understand why a segfault occurs only for the SSE2-optimized Atlas library (and not for reference blas, or Atlas with other optimizations).

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2008-04-12

SOLVED: this is an aligment issue (see below for proof).

Apparently, something in the SSE2 code in Atlas requires that incoming doubles are aligned to a 8-byte boundary and causes a crash if they are not. This is surprising, as documentation on the net says that SSE2 can optionally assume that values lie at 16-byte boundaries! However, GNU libc docs state that "In the GNU system, the address is always a multiple of eight on most systems, and a multiple of 16 on 64-bit systems.", so maybe there is some broken optimization done based on this assumption. (And apparently the byte data in Python strings is consistently off by 4 from 8-byte alignment.)

So Atlas + SSE2 is somehow broken (at least, unless this is clearly explained in its docs...). The question is what numpy should do about this.

---

Finally, to believe that it's really the alignment at fault here, apply the patch
    diff -r d6b1b710efac numpy/core/src/arraymethods.c
    --- a/numpy/core/src/arraymethods.c Sat Apr 12 06:12:09 2008 +0300
    +++ b/numpy/core/src/arraymethods.c Sun Apr 13 02:04:58 2008 +0300
    atmention:atmention: -1229,6 +1229,16 atmention:atmention: array_setstate(PyArrayObject *self, PyOb
             if (PyString_AsStringAndSize(rawdata, &datastr, &len))
                 return NULL;

```
+        {
+            char *dupmem;
+            dupmem = malloc(len + 16);
+            dupmem += (((long long)dupmem) % 8) + 0; /* doesn't crash */
+            /* dupmem += (((long long)dupmem) % 8) + 4;*/ /* crashes */
+            memcpy(dupmem, datastr, len);
+            datastr = dupmem;
+            printf("--xx-- %p, %lld\n", dupmem, ((long long)dupmem) % 16);
+        }
+
         if ((len != (self->descr->elsize * size))) {
             PyErr_SetString(PyExc_ValueError,
                             "buffer size does not"  \
```

and run the following test code, using first numpy compiled with the first marked line commented out, and then the second line commented out:
    $ cat test.py
    import numpy

```
x = numpy.zeros((3,4))
v = numpy.ones((1, x.shape[0]))

z = x.__reduce__()[2]
y = numpy.empty(x.shape, x.dtype)
y.__setstate__(z)

print "x:", x.__array_interface__['data'][0] % 16
print "y:", y.__array_interface__['data'][0] % 16

print "1"
numpy.dot(v, x)
print "2"
numpy.dot(v, y)
print "3"
$ python test.py   # first run
--xx-- 0x8327128, 8
x: 8
y: 8
1
2
3
$ comment out the second line, comment the first line and recompile numpy
$ python test.py
--xx-- 0x832712c, 12
x: 8
y: 12
1
2
Segmentation fault
```

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2008-04-13

... and a minimal testcase:
    $ cat sse2-crash-test.c
    #include <stdlib.h>
    #include <string.h>
    #include <stdio.h>

```
#include "cblas.h"

void *aligned_malloc(unsigned char align, unsigned char offset, int size)
{
    char *ptr;
    ptr = malloc(size + offset + align);
    memset(ptr, 0, size + offset + align);
    ptr += (((long)ptr) % align) + offset;
    return (void*)ptr;
}

int main() {
    int m = 3, n = 4;
    double *good_a, *bad_a, *b, *c;

    good_a = (double*)aligned_malloc(8, 0, m*n*sizeof(double)); /* m x n */
    bad_a  = (double*)aligned_malloc(8, 4, m*n*sizeof(double)); /* m x n */
    b      = (double*)aligned_malloc(8, 0, m*sizeof(double));   /* m x 1 */
    c      = (double*)aligned_malloc(8, 0, n*sizeof(double));   /* n x 1 */

    /* c := a^T b */

    fprintf(stderr, "good_align: ");
    cblas_dgemv(CblasRowMajor,CblasTrans,m,n, 1.0, good_a, n, b, 1, 0.0, c, 1);
    fprintf(stderr, "ok\n");

    fprintf(stderr, "bad_align: ");
    cblas_dgemv(CblasRowMajor,CblasTrans,m,n, 1.0, bad_a, n, b, 1, 0.0, c, 1);
    fprintf(stderr, "ok\n");

    return 0;
}

$ dpkg --list|grep -E 'atlas3.*(sse2|base)'
ii  atlas3-base     3.6.0-20.6     Automatically Tuned Linear Algebra Software,
ii  atlas3-base-dev 3.6.0-20.6     Automatically Tuned Linear Algebra Software,
ii  atlas3-sse2     3.6.0-20.6     Automatically Tuned Linear Algebra Software,
ii  atlas3-sse2-dev 3.6.0-20.6     Automatically Tuned Linear Algebra Software,

$ gcc -o sse2-crash-test sse2-crash-test.c -lcblas -lm
$ LD_PRELOAD=/usr/lib/sse2/libcblas.so.3.0 ./sse2-crash-test
good_align: ok
bad_align: Segmentation fault
$ LD_PRELOAD=/usr/lib/libcblas.so.3.0 ./sse2-crash-test
good_align: ok
bad_align: ok
```

Can someone else confirm this?

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2008-04-13

This certainly looks to me that alignment is the issue.   Very good sleuthing.   Can we check for this particular configuration and copy if not aligned in such cases.  Or should we just copy to 16-byte aligned all the time before handing off to vendor BLAS.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2008-04-13

Since creating arrays from user supplied memory is a feature, maybe we should check and return an error if the supplied memory is not aligned. For our own use we have David's allocator.

```
ptr += (((long)ptr) % align) + offset;
```

That doesn't look right.

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2008-04-13

The user supplied memory here typically comes from python.pickle, which the end-user has no control over, so I don't think raising an exception is a good idea for this case.

And yes, do replace "long" by "unsigned long" in the codes, 
    ptr += (((unsigned long)ptr) % align) + offset;
if you meant this, in C % can return negative numbers... (And you can't free(ptr) after this, so even this is good only for testing...) Anyway, this didn't change the results.

**Comment by thouis** [comment]
Comment in Trac by atmention:rkern, 2008-04-13

I should think size_t would be the appropriate integer type rather than unsigned long. longs are 32-bit on Windows 64-bit.

**Comment by thouis** [comment]
Comment in Trac by trac user peridot, 2008-05-06

dot() and friends can't afford to assume that arrays have any particular alignment, since users can always fabricate arrays with arbitrary alignments. At the least dot() must detect this and raise an exception rather than segfaulting. How hard would it be to fall back to generic BLAS if the arrays are not aligned? Copying works too, but copying can be very expensive, memory-wise. Imagine a user who uses a giant matrix two-thirds the size of available RAM to dot() against many vectors. Every time she calls dot(), a copy is made, one copy is swapped out to disk, the dot() is done, and the new matrix is thrown away. Worse, imagine she tests this on her laptop, which has generic BLAS, then tries to move it to the cluster nodes, which have ATLAS and suddenly require this copy. Suddenly a massive, hard-to-debug, slowdown.

Of course, having arrays be aligned whenever possible will help, but it's unreasonable to enforce alignment.

**Comment by thouis** [comment]
Comment in Trac by trac user peridot, 2008-05-06

Does this bug affect record arrays as well? 

For example:

In [9]: A = np.zeros(10,dtype=[('a',np.double),('b',np.int32)])

Every other float is four bytes off.

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2008-05-07

Replying to [comment:28 peridot]:

> Does this bug affect record arrays as well? 
> 
> For example:
>     In [9]: A = np.zeros(10,dtype=[('a',np.double),('b',np.int32)])
> Every other float is four bytes off.

dot (the A.v DGEMV path that is cause of problems here, _dotblas.c:550) makes a copy if the array is not one-segment, so the bug doesn't affect this particular example.

**Comment by thouis** [comment]
Comment in Trac by trac user peridot, 2008-05-07

Replying to [comment:29 pv]:

> Replying to [comment:28 peridot]:
> 
> > Does this bug affect record arrays as well? 
> > 
> > For example:
> > {{{
> > In [9]: A = np.zeros(10,dtype=[('a',np.double),('b',np.int32)])
> > }}}
> > Every other float is four bytes off.
> 
> dot (the A.v DGEMV path that is cause of problems here, _dotblas.c:550) makes a copy if the array is not one-segment, so the bug doesn't affect this particular example.

Ah. Then maybe it does make sense for it to make a copy for misaligned arrays as well. Seems like it would solve some very puzzling errors.

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2008-05-07

This should be fixed in r5141 (for 1.1.x) and in r5140 (for 1.2).

Basically, the code now copies data if the start of the input is not aligned on an itemsize boundary (8-bytes for doubles and 4-bytes for floats).

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2008-05-08

Thanks, the fix works.

Tests added in r5147 (trunk) and in r5148 (1.1.x).
